Add policy monitor debounce time

michel-laterman opened the pull request:
## What is the problem this PR solves?

scale tests frequently fail on the policy change tests where fleet-server has not finished dispatching policy change actions before a new change is detected.

## How does this PR solve the problem?

Add a policy_debounce_time configuration to add a forced delay to the
policy index monitor when it successfully gathers new documents. This
is needed to stop fleet-server from dispatching all policies in cases
where policy changes occur frequently and fleet-server is unable to
dispatch all policy change actions before the policy is updated.

## Design Checklist

- [x] I have ensured my design is stateless and will work when multiple fleet-server instances are behind a load balancer.
- [ ] I have or intend to scale test my changes, ensuring it will work reliably with 100K+ agents connected.
- [x] I have included fail safe mechanisms to limit the load on fleet-server: rate limiting, circuit breakers, caching, load shedding, etc.

## Checklist

- [x] I have commented my code, particularly in hard-to-understand areas
- ~~I have made corresponding changes to the documentation~~
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/fleet-server#changelog)

michel-laterman commented:
the integration tests don't get counted as far as code coverage goes (i'm not sure why), so sonarqube will mark this as a failure even though a test has been added.

michel-laterman commented:
buildkite run perf-tests

juliaElastic commented on internal/pkg/monitor/monitor_integration_test.go:
```diff
@@ -47,6 +48,66 @@ func TestSimpleMonitorNonEmptyIndex(t *testing.T) {
+	defer cn()
+	ctx = testlog.SetLogger(t).WithContext(ctx)
+
+	index, bulker := ftesting.SetupCleanIndex(ctx, t, ".fleet-actions")
```
Shouldn't the test add new docs in `.fleet-policies` index to be more realistic? Policy changes don't have an action document, I think.

juliaElastic commented:
I've run a 5k test (with removing the sleep in multiple policy change test) with a custom image from this pr. The multiple policy changes succeeded quickly, which is a good sign:
https://buildkite.com/elastic/observability-perf/builds/2220#018d3b16-3dd2-41e9-8f45-b042f6e8aba5
```

╭─ Step05: test step 5: AC-21 - Apply multiple policy changes without waiting for prior to complete ───────────────────────────────────────╮
--
  | │                                                                                                                                          │
  | │  Multiple back-to-back policy changes.                                                                                                   │
  | │                                                                                                                                          │
  | │      - Make N changes to nginx integration of a policy                                                                                   │
  | │      - Agent policy revision should be bumped by N                                                                                       │
  | │      - All agents should remain online                                                                                                   │
  | │                                                                                                                                          │
  | │      Has to be run in conjunction with Step04, because nginx integration is added there                                                  │
  | │                                                                                                                                          │
  | ╰──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
  | [11:07:30] INFO     oblt-perf.perf_lib: "nginx" integration was successfully updated
  | [11:07:34] INFO     oblt-perf.perf_lib: "nginx" integration was successfully updated
  | [11:07:37] INFO     oblt-perf.perf_lib: "nginx" integration was successfully updated
  | [11:07:40] INFO     oblt-perf.perf_lib: "nginx" integration was successfully updated
  | [11:07:44] INFO     oblt-perf.perf_lib: "nginx" integration was successfully updated
  | [11:07:44] INFO     oblt-perf.perf_lib: Looking for "online" greaterThanOrEqualTo 5000
  | [11:07:44] INFO     oblt-perf.perf_lib: Timemout in 20m00s
  | [11:07:44] INFO     oblt-perf.perf_lib: poll_agents_status kuery is "fleet-agents.policy_id : 782ac8d3-ac1c-4518-a370-6f52e52678dc and
  | policy_revision_idx >= 8"
  | [11:07:44] INFO     oblt-perf.perf_lib: {'total': 0, 'inactive': 0, 'online': 0, 'error': 0, 'offline': 0, 'updating': 0, 'other': 0,
  | 'events': 0, 'unenrolled': 0, 'all': 0, 'active': 0} undetermined
  | [11:08:14] INFO     oblt-perf.perf_lib: {'total': 0, 'inactive': 0, 'online': 0, 'error': 0, 'offline': 0, 'updating': 0, 'other': 0,
  | 'events': 0, 'unenrolled': 0, 'all': 0, 'active': 0} undetermined
  | [11:08:44] INFO     oblt-perf.perf_lib: {'total': 0, 'inactive': 0, 'online': 0, 'error': 0, 'offline': 0, 'updating': 0, 'other': 0,
  | 'events': 0, 'unenrolled': 0, 'all': 0, 'active': 0} not ascending
  | [11:09:15] INFO     oblt-perf.perf_lib: {'total': 3894, 'inactive': 0, 'online': 3894, 'error': 0, 'offline': 0, 'updating': 0, 'other': 0,
  | 'events': 0, 'unenrolled': 0, 'all': 3894, 'active': 3894} not ascending
  | [11:09:45] INFO     oblt-perf.perf_lib: {'total': 4482, 'inactive': 0, 'online': 4482, 'error': 0, 'offline': 0, 'updating': 0, 'other': 0,
  | 'events': 0, 'unenrolled': 0, 'all': 4482, 'active': 4482} ascending
  | [11:10:15] INFO     oblt-perf.perf_lib: {'total': 4990, 'inactive': 0, 'online': 4990, 'error': 0, 'offline': 0, 'updating': 0, 'other': 0,
  | 'events': 0, 'unenrolled': 0, 'all': 4990, 'active': 4990} ascending
  | [11:10:46] INFO     oblt-perf.perf_lib: {'total': 5000, 'inactive': 0, 'online': 5000, 'error': 0, 'offline': 0, 'updating': 0, 'other': 0,
  | 'events': 0, 'unenrolled': 0, 'all': 5000, 'active': 5000} ascending
  | [11:10:46] INFO     oblt-perf.perf_lib: "online" 5000/5000 (current/desired) 2m32s
  | [11:10:46] INFO     oblt-perf: FleetAgentStatus(total=5000, inactive=0, online=5000, error=0, offline=0, updating=0, other=0, events=0,
  | unenrolled=0, all=5000, active=5000, doc_id=None, run_id=None, timestamp=None, kuery='fleet-agents.policy_id :
  | 782ac8d3-ac1c-4518-a370-6f52e52678dc and policy_revision_idx >= 8', cluster_name=None)
  | ───────────────────────────────────── harness_0_step_5_iteration_1 :: COMPLETE :: Elapsed time: 3m22s ───
```
In horde logs I see the policy change revision 4 and 8 was received, the revisions in between were skipped, which is good.
What I don't understand, why the revision 8 is sent twice, and the next step with revision 9 doesn't seem to be propagated to all agents.
<img width="764" alt="image" src="https://github.com/elastic/fleet-server/assets/90178898/bdf33f49-3ca6-4482-b994-2107f08258ea">

```

─────────────────────────────────────── harness_0_step_6_iteration_1 :: STARTING :: Elapsed time: 0 ────────────────────────────────────────
--
  | ╭─ Step06: test step 6: AC-9 - Remove integration from policy ─────────────────────────────────────────────────────────────────────────────╮
  | │                                                                                                                                          │
  | │  Deleting integration from a policy that is already in use by multiple agents.                                                           │
  | │                                                                                                                                          │
  | │      - Nginx integration is expected to be removed from the policy                                                                       │
  | │      - Agent policy revision to be bumped                                                                                                │
  | │      - All selected agents should remain "Healthy" (online).                                                                             │
  | ╰──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
  | [11:10:47] INFO     oblt-perf.perf_lib: Updated list of integrations in swarm_0-pp1-ecf328 at rev 8: ['system', 'elastic_agent', 'endpoint',
  | 'nginx']
  | [11:10:51] INFO     oblt-perf: Checking if the changes were applied to all agents...
  | [11:10:51] INFO     oblt-perf.perf_lib: Looking for "online" greaterThanOrEqualTo 5000
  | [11:10:51] INFO     oblt-perf.perf_lib: Timemout in 20m00s
  | [11:10:51] INFO     oblt-perf.perf_lib: poll_agents_status kuery is "fleet-agents.policy_id : 782ac8d3-ac1c-4518-a370-6f52e52678dc and
  | policy_revision_idx >= 9"
  | [11:10:51] INFO     oblt-perf.perf_lib: {'total': 0, 'inactive': 0, 'online': 0, 'error': 0, 'offline': 0, 'updating': 0, 'other': 0,
  | 'events': 0, 'unenrolled': 0, 'all': 0, 'active': 0} undetermined
  | [11:11:21] INFO     oblt-perf.perf_lib: {'total': 533, 'inactive': 0, 'online': 533, 'error': 0, 'offline': 0, 'updating': 0, 'other': 0,
  | 'events': 0, 'unenrolled': 0, 'all': 533, 'active': 533} undetermined
  | [11:11:51] INFO     oblt-perf.perf_lib: {'total': 1427, 'inactive': 0, 'online': 1427, 'error': 0, 'offline': 0, 'updating': 0, 'other': 0,
  | 'events': 0, 'unenrolled': 0, 'all': 1427, 'active': 1427} ascending
  | [11:12:22] INFO     oblt-perf.perf_lib: {'total': 2180, 'inactive': 0, 'online': 2180, 'error': 0, 'offline': 0, 'updating': 0, 'other': 0,
  | 'events': 0, 'unenrolled': 0, 'all': 2180, 'active': 2180} ascending
  | [11:12:52] INFO     oblt-perf.perf_lib: {'total': 2180, 'inactive': 0, 'online': 2180, 'error': 0, 'offline': 0, 'updating': 0, 'other': 0,
  | 'events': 0, 'unenrolled': 0, 'all': 2180, 'active': 2180} ascending

```

juliaElastic commented:
I've run another test with 10k and 10 subsequent policy changes without sleep, it timed out at 20m at 6k
https://buildkite.com/elastic/observability-perf/builds/2222#018d3b4f-8a23-4c7c-9270-3b4e59259bee

michel-laterman commented on internal/pkg/policy/monitor_integration_test.go:
```diff
@@ -128,3 +129,162 @@ func TestMonitor_Integration(t *testing.T) {
+
+	index, bulker := ftesting.SetupCleanIndex(ctx, t, dl.FleetPolicies)
+
+	im, err := monitor.New(index, bulker.Client(), bulker.Client(), monitor.WithDebounceTime(time.Second))
```
@juliaElastic, I've added another integration test to test the `policy.Monitor` when the underlying index-monitor has the debounce time set.

This test explicitly makes a couple updates within the debounce time to ensure that only the latest is sent to a subscription

michel-laterman commented:
buildkite run perf-tests

michel-laterman commented:
Not sure why the perf-tests succeeded the 2nd attempt I just did, but failed the previous attempt.

juliaElastic commented:
I've run another scale test with the latest changes, and the multiple policy change slowed down after 5m around 3k of our 5k: https://buildkite.com/elastic/observability-perf/builds/2231#018d400e-01d7-49d6-bfdc-3997aa6b1a86
I stopped the execution to see if the drones eventually complete, all drones seem online and checking in regularly, but some of them haven't received the latest revision yet (revision changed from 3 to 8 quickly).
Interestingly none of the drones stayed at revision 3, but about 2k is stuck at revision 4.

michel-laterman commented:
That message indicates that it took fleet-server a long time to authorize the API key for the checkin

juliaElastic commented:
I ran another test with 15s debounce time (so it encompasses the 5 subsequent policy changes), and this time the testcase passed: https://buildkite.com/elastic/observability-perf/builds/2235#018d4149-ad43-417f-8c26-6efd5e349b90

So I think something is wrong when there is a policy change after the debounce time expires, the monitor should pick up the new revision.

```

[15:54:07] INFO     oblt-perf.perf_lib: "nginx" integration was successfully updated
--
  | [15:54:09] INFO     oblt-perf.perf_lib: "nginx" integration was successfully updated
  | [15:54:12] INFO     oblt-perf.perf_lib: "nginx" integration was successfully updated
  | [15:54:15] INFO     oblt-perf.perf_lib: "nginx" integration was successfully updated
  | [15:54:17] INFO     oblt-perf.perf_lib: "nginx" integration was successfully updated
  | [15:54:18] INFO     oblt-perf.perf_lib: Looking for "online" greaterThanOrEqualTo 5000
  | [15:54:18] INFO     oblt-perf.perf_lib: Timemout in 20m00s
  | [15:54:18] INFO     oblt-perf.perf_lib: poll_agents_status kuery is "fleet-agents.policy_id : ac2a72a3-4bcc-414f-a8ff-b8f85019f83b and
  | policy_revision_idx >= 8"
  | [15:54:18] INFO     oblt-perf.perf_lib: {'total': 0, 'inactive': 0, 'online': 0, 'error': 0, 'offline': 0, 'updating': 0, 'other': 0,
  | 'events': 0, 'unenrolled': 0, 'all': 0, 'active': 0} undetermined
  | [15:54:48] INFO     oblt-perf.perf_lib: {'total': 0, 'inactive': 0, 'online': 0, 'error': 0, 'offline': 0, 'updating': 0, 'other': 0,
  | 'events': 0, 'unenrolled': 0, 'all': 0, 'active': 0} undetermined
  | [15:55:18] INFO     oblt-perf.perf_lib: {'total': 0, 'inactive': 0, 'online': 0, 'error': 0, 'offline': 0, 'updating': 0, 'other': 0,
  | 'events': 0, 'unenrolled': 0, 'all': 0, 'active': 0} not ascending
  | [15:55:49] INFO     oblt-perf.perf_lib: {'total': 0, 'inactive': 0, 'online': 0, 'error': 0, 'offline': 0, 'updating': 0, 'other': 0,
  | 'events': 0, 'unenrolled': 0, 'all': 0, 'active': 0} not ascending
  | [15:56:19] INFO     oblt-perf.perf_lib: {'total': 401, 'inactive': 0, 'online': 401, 'error': 0, 'offline': 0, 'updating': 0, 'other': 0,
  | 'events': 0, 'unenrolled': 0, 'all': 401, 'active': 401} not ascending
  | [15:56:49] INFO     oblt-perf.perf_lib: {'total': 2106, 'inactive': 0, 'online': 2106, 'error': 0, 'offline': 0, 'updating': 0, 'other': 0,
  | 'events': 0, 'unenrolled': 0, 'all': 2106, 'active': 2106} ascending
  | [15:57:19] INFO     oblt-perf.perf_lib: {'total': 2603, 'inactive': 0, 'online': 2603, 'error': 0, 'offline': 0, 'updating': 0, 'other': 0,
  | 'events': 0, 'unenrolled': 0, 'all': 2603, 'active': 2603} ascending
  | [15:57:50] INFO     oblt-perf.perf_lib: {'total': 4963, 'inactive': 0, 'online': 4963, 'error': 0, 'offline': 0, 'updating': 0, 'other': 0,
  | 'events': 0, 'unenrolled': 0, 'all': 4963, 'active': 4963} ascending
  | [15:58:20] INFO     oblt-perf.perf_lib: {'total': 4993, 'inactive': 0, 'online': 4993, 'error': 0, 'offline': 0, 'updating': 0, 'other': 0,
  | 'events': 0, 'unenrolled': 0, 'all': 4993, 'active': 4993} ascending
  | [15:58:50] INFO     oblt-perf.perf_lib: {'total': 5000, 'inactive': 0, 'online': 5000, 'error': 0, 'offline': 0, 'updating': 0, 'other': 0,
  | 'events': 0, 'unenrolled': 0, 'all': 5000, 'active': 5000} ascending
  | [15:58:50] INFO     oblt-perf.perf_lib: "online" 5000/5000 (current/desired) 4m03s
  | [15:58:51] INFO     oblt-perf: FleetAgentStatus(total=5000, inactive=0, online=5000, error=0, offline=0, updating=0, other=0, events=0,
  | unenrolled=0, all=5000, active=5000, doc_id=None, run_id=None, timestamp=None, kuery='fleet-agents.policy_id :
  | ac2a72a3-4bcc-414f-a8ff-b8f85019f83b and policy_revision_idx >= 8', cluster_name=None)
  | ───────────────────────────────────── harness_0_step_5_iteration_1 :: COMPLETE :: Elapsed time: 4m49s


```

juliaElastic commented:
I couldn't reproduce the issue with 50 drones, so it might occur only on scale: https://buildkite.com/elastic/observability-perf/builds/2247#018d4561-9473-419c-9930-eeb9694f422f
In this build the debounce time is 1s, and I see the revisions 4,5,6,7,8 dispatched, and all drones update to revision 8.

michel-laterman commented:
The dropped notification messages are an issue, they mean that the update was dropped because it timed out while writing to the subscription.
Can we get debug logs for these? I would like to see (on one of the failing agents) what the checkin revision idx is when it fails

juliaElastic commented:
> The dropped notification messages are an issue, they mean that the update was dropped because it timed out while writing to the subscription. Can we get debug logs for these? I would like to see (on one of the failing agents) what the checkin revision idx is when it fails

The debug logs are under the same link (just filtered them out as there were many of them) if you mean the fleet-server logs.
https://overview.elastic-cloud.com/app/r/s/dNj74

michel-laterman commented:
The revision numbers we log are associated with two different things.
The dispatch message's revision number is associated with the subscription (`s.revIdx`), or the agent's rev at time of checkin:
https://github.com/elastic/fleet-server/blob/5e8a1e9f9f3cf9d7b97e8c41a7c6611b414eca55/internal/pkg/policy/monitor.go#L224-L230

The updatePolicies method uses the one associated with the policy:
https://github.com/elastic/fleet-server/blob/5e8a1e9f9f3cf9d7b97e8c41a7c6611b414eca55/internal/pkg/policy/monitor.go#L309-L313

I've added it so make it a little more clear in test case:
```
    monitor_integration_test.go:236: added rev 4 policy to index
    monitor.go:192: {"level":"debug","index":".fleet-policies","ctx":"index monitor","checkpoints":[1549],"message":"updated checkpoint"}
    monitor.go:341: {"level":"debug","index":".fleet-policies","ctx":"index monitor","count":1,"message":"hits found after notify"}
    monitor.go:192: {"level":"debug","index":".fleet-policies","ctx":"index monitor","checkpoints":[1549],"message":"updated checkpoint"}
    monitor.go:364: {"level":"debug","ctx":"policy agent monitor","fleet.policy.id":"aec60adf-fab4-4433-bcbd-731c0ec2de35","rev":4,"coord":1,"fleet.agent.id":"6260d564-7b6b-4f6c-b345-7e3f78291cbd","sub_revidx":3,"message":"scheduled pendingQ on policy revision"}
    monitor.go:375: {"level":"info","ctx":"policy agent monitor","fleet.policy.id":"aec60adf-fab4-4433-bcbd-731c0ec2de35","rev":4,"coord":1,"oldRev":3,"oldCoord":1,"nQueued":1,"fleet.policy.id":"aec60adf-fab4-4433-bcbd-731c0ec2de35","message":"New revision of policy received and added to the queue"}
    monitor_integration_test.go:295: received third policy from subsciption, rev 4
    monitor.go:230: {"level":"debug","ctx":"policy agent monitor","fleet.agent.id":"6260d564-7b6b-4f6c-b345-7e3f78291cbd","fleet.policy.id":"aec60adf-fab4-4433-bcbd-731c0ec2de35","rev":3,"coord":1,"message":"dispatch"}
```
In the last line it looks like rev 3 was sent, but but the line above it indicates that rev 4 was sent to the policy monitor's output channel.

juliaElastic commented on internal/pkg/policy/monitor.go:
```diff
@@ -225,8 +225,10 @@ func (m *monitorT) dispatchPending(ctx context.Context) {
-				Int64("coord", s.coordIdx).
+				Int64("sub_rev", s.revIdx).
+				Int64("sub_coord", s.coordIdx).
+				Int64("rev", &policy.pp.Policy.RevisionIdx).
```
we should make the log fields consistent, I've seen revision being logged out as `revision_idx` somewhere else
```
Policy coordinator added a new policy revision
revision_idx
```

Similarly, the action type if logged under different fields:
```
ack event
actionId

Action delivered to agent on checkin
id
```

michel-laterman replied:
I think i got most of the revision/coordinator attributes to be consistent, but i've made a follow up to generally go through the codebase to check: https://github.com/elastic/fleet-server/issues/3239

juliaElastic commented:
@michel-laterman What can we do to fix this? Should there be a retry for the "dropped notification" error?

michel-laterman commented:
> @michel-laterman What can we do to fix this? Should there be a retry for the "dropped notification" error?

I don't think that's a good idea, dropping in this case indicates high load. If we're able to collect profiles/a diagnostics bundle from an instance where this occurs I think we can investigate this more

michel-laterman commented:
buildkite run perf-tests

juliaElastic commented:
I was thinking that while we investigate the dropped notifications issue, we could get this change in and use a longer debounce for scale tests to prevent the issue (15s would fix the 5 subsequent policy changes).

juliaElastic commented:
@blakerouse I wonder if you have any tips here, since you worked on the policy monitor a few years back.

michel-laterman commented:
buildkite run perf-tests

juliaElastic commented:
I've tried another [test](https://buildkite.com/elastic/observability-perf/builds/2284#018d5f0f-b7ab-4d71-a325-bd03bb6df893) with 5k with `memory_limit` [set to 2gb](https://github.com/elastic/observability-perf/pull/724/commits/a2ea385de5fc6adfc2d72b079b50cc42c2279a2e) on fleet-server (the default was 4gb). 
This seems to have helped, the multiple policy change succeeded. It seems that the issue is related to fleet-server being low on memory (though I don't see any indication of this on the metrics dashboard, it shows about 700 MB usage).

juliaElastic commented on internal/pkg/policy/monitor.go:
```diff
@@ -303,13 +343,19 @@ func (m *monitorT) groupByLatest(policies []model.Policy) map[string]model.Polic
 		Str(logger.PolicyID, newPolicy.PolicyID).
-		Int64("rev", newPolicy.RevisionIdx).
-		Int64("coord", newPolicy.CoordinatorIdx).
+		Int64("revision_idx", newPolicy.RevisionIdx).
```
we could refer the constant `dl.FieldRevisionIdx` here

juliaElastic commented on internal/pkg/policy/monitor_integration_test.go:
```diff
@@ -128,3 +129,410 @@ func TestMonitor_Integration(t *testing.T) {
+
+	policy := model.Policy{
+		PolicyID:       policyID,
+		CoordinatorIdx: 1,
```
Could we have a test which tests the coordinator logic too? And adds a new policy revision with `coordinator_idx:0`

juliaElastic commented:
I tested with more logging and attempted to fix, but something is still not right.
I noticed that sometimes it takes a long time for the policy change to be dispatched on the agent subscription channels (30s-1m for 5k), I thought that is because the channel buffer size was 1, so in case of multiple dispatches the code blocks until the agent checks in to read the channel. Increased the buffer size to 5 in this [commit](https://github.com/elastic/fleet-server/pull/3234/commits/fae415b0719fd0319365219a0be9b30277d3b342), but still some revisions don't seem to be dispatched to some agents.

latest run with the changes in this pr: https://buildkite.com/elastic/observability-perf/builds/2315

<img width="1444" alt="image" src="https://github.com/elastic/fleet-server/assets/90178898/11374483-07ac-453b-b6d3-7eddbbb8883b">

What's also weird in this run, is that there seems to be a `revision_idx:7` dispatched after `revision_idx:8` by the second instance.

<img width="1486" alt="image" src="https://github.com/elastic/fleet-server/assets/90178898/6c71e547-5f9a-47f3-81e0-dc1ba9ec38f3">

juliaElastic commented:
I noticed that before a `dropped notification` happens, there are many `authApiKey slow` messages. It could be related why the timeout in the subscription monitor is reached.

<img width="2264" alt="image" src="https://github.com/elastic/fleet-server/assets/90178898/8f3151f3-8f95-4b01-bd50-54ba358a18d2">

michel-laterman commented:
OK, i'm going to try to reply to a lot of things here

> Also noticed that on agent checkin, the agent subscribes to the policy monitor with the agent's current revision. I'm not sure though why some agents don't get the revision 8.
Yes, that's expected. When an agent checks in fleet-server gets the current revision number the agent is running by reading the revision_idx of the agent doc from the .fleet-agents index.
The subscription revision is used to only send a policy if the policy_monitor detects a document with a higher revision_idx then the subscription.

> I think we should have some kind of fallback mechanism to read the latest policy revision in case the notification is dropped.
[groupByLatest](https://github.com/elastic/fleet-server/blob/main/internal/pkg/policy/monitor.go#L284) should be doing this

> Increased the buffer size to 5 in this [commit](https://github.com/elastic/fleet-server/pull/3234/commits/fae415b0719fd0319365219a0be9b30277d3b342), but still some revisions don't seem to be dispatched to some agents.
I think that a policy monitor subscription is only expected to produce a single output.
This might be a cause of the issue.

The 1st policy will be processed by [updatePolicy](https://github.com/elastic/fleet-server/blob/main/internal/pkg/policy/monitor.go#L306) and dispatched to any subscribed agents (5-10k in our scale tests), you said earlier it takes 30s-1m for 5k agents, the long time may mean that the policy monitor is going to process all changes, and dispatch them.
I think that to really solve the issue we would need to update/dispatch in a different goroutine then the main policy monitor run loop, and that we should be able to cancel/interrupt an update/dispatch if a new policy document is received.
However, this will be a lot of work and should probably be split into another issue.

I think your proposed changes, increase/remove the timeout on subscription dispatches, and increase the debounce time should work for the time being.

aleksmaus commented:
> @aleksmaus Hey, do you remember why the 5s timeout was introduced [here](https://github.com/elastic/fleet-server/pull/46/files#diff-d8026fd248736718d440cf6a898c5c3f3cd3d140a6b7df81cd5ece9f46fd9830)? It seems we have an issue when it takes longer to pick up the change, and the notification is dropped, and never re-sent.

@blakerouse do you remember what was that timeout about?

blakerouse commented:
@aleksmaus That is not something that was changed in the refactor. It was always 5 seconds - https://github.com/elastic/fleet-server/pull/46/files#diff-3e9fdf7b7163c16024ec81594416a56bd4706b609fd175664f7f7a20399cde89L26

I think the idea of having a timeout was always just to ensure that a deadlock didn't occur, I don't think we really set any really reason for the amount of time.

juliaElastic commented:
> I think the idea of having a timeout was always just to ensure that a deadlock didn't occur, I don't think we really set any really reason for the amount of time.

Thanks for the explanation. I'm wondering how likely it is to hit a deadlock, so far I haven't seen any negative impact of removing the timeout.

juliaElastic commented:
@michel-laterman Any ideas why the integration test step is failing? I don't see any failing tests.
